### PR TITLE
#15844 adding language support and refactor

### PR DESCRIPTION
**Pull request by jgambarios**

### CODE REVIEW

**Comment by jgambarios** [comment]
@jdotcms I reverted the original pull request and I open another one in order to apply the feedback from Will https://github.com/dotCMS/core/pull/15851#pullrequestreview-193888064

**Comment by wezell** [comment]
Guys, I added on to this branch so the language would respect the fallback appropriately.  @dsilvam I added this method though I would take any suggestions for better names:

https://github.com/dotCMS/core/pull/15854/files#diff-0e42bc8d57cb229663e56b48259b767dR343

Let's talk on Tuesday.

`findContentletByIdentifierAndLangOrFallback`?

File: dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java [hunk 1]
```diff
@@ -12,177 +12,279 @@
...
 
+    return width;
+  }
 
+  private int getHeight (final String uri, final int defaultHeight) {
 
-    // Checking if host is active
-    if (!mode.isAdmin && !APILocator.getVersionableAPI().hasLiveVersion(host)) {
-      response.sendError(HttpServletResponse.SC_SERVICE_UNAVAILABLE,
-          LanguageUtil.get("server-unavailable-error-message"));
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java]: It looks like we lost the host.isLive() check - we should bring this logic back as it is valid

Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java]: It has been moved to the method isValidRequest

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java [hunk 2]
```diff
@@ -336,6 +337,39 @@ public Contentlet findContentletByIdentifier(String identifier, boolean live, lo
 
     }
 
+    
+    @CloseDBIfOpened
+    @Override
+    public Optional<Contentlet> findContentletByIdentifierOrFallback(final String identifier, final boolean live,
```
Comment (jgambarios) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java]: As we discussed we need integration tests for this method.

File: dotCMS/src/main/java/com/dotcms/contenttype/model/type/ContentType.java [hunk 3]
```diff
@@ -266,4 +267,25 @@ public String getPermissionType() {
     return ImmutableList.of();
   }
 
+  @JsonIgnore
+  @Value.Lazy
+  public boolean languageFallback() {
```
Comment (jgambarios) [inline dotCMS/src/main/java/com/dotcms/contenttype/model/type/ContentType.java]: As we discussed we need integration tests for this method.

File: dotCMS/src/main/java/com/dotcms/contenttype/model/type/ContentType.java [hunk 4]
```diff
@@ -266,4 +267,25 @@ public String getPermissionType() {
...
+    return baseType() == BaseContentType.CONTENT
+            ? Config.getBooleanProperty("DEFAULT_CONTENT_TO_DEFAULT_LANGUAGE",false)
+                : baseType() == BaseContentType.WIDGET
+                ? Config.getBooleanProperty("DEFAULT_WIDGET_TO_DEFAULT_LANGUAGE",false)
+                        : baseType() == BaseContentType.FILEASSET
+                                ? Config.getBooleanProperty("DEFAULT_FILE_TO_DEFAULT_LANGUAGE",false)
+                                        : baseType() == BaseContentType.FILEASSET
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/contenttype/model/type/ContentType.java]: done

File: dotCMS/src/main/java/com/dotcms/util/ConversionUtils.java [hunk 5]
```diff
@@ -133,6 +137,37 @@ public static long toLong (final Object input, final Long defaultLong) {
...
+	 * can be a String or an instance of {@link Number}.
+	 * @param input
+	 *         - The value to convert.
+	 * @param defaultLong
+	 *        - Supplier with the default value in case the input cannot be converted.
+	 * @return long value
+	 */
+	public static long toLong (final Object input, final Supplier<Long> defaultLong) {
+
+		long l = 0;
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/util/ConversionUtils.java]: done

File: dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java [hunk 6]
```diff
@@ -12,177 +12,279 @@
...
+ */
 public class ShortyServlet extends HttpServlet {
 
   private static final long serialVersionUID = 1L;
 
-  final static String FILE_ASSEST_DEFAULT = FileAssetAPI.BINARY_FIELD;
+  private final HostWebAPI     hostWebAPI     = WebAPILocator.getHostWebAPI();
+  private final VersionableAPI versionableAPI = APILocator.getVersionableAPI();
+  private final ShortyIdAPI    shortyIdAPI    = APILocator.getShortyAPI();
+  private final ContentletAPI  contentletAPI  = APILocator.getContentletAPI();
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/servlets/ShortyServlet.java]: done

**Pull request merged**